### PR TITLE
Replace scale down window

### PR DESCRIPTION
**Pull request by krzysztof-jastrzebski**
**What this PR does / why we need it**:
Replace scale down forbidden window with scale down stabilization window.

This allows scale down based on more than one sample, to avoid rapidly changing size up and down for controllers with fluctuating load.

A bit more in https://docs.google.com/document/d/1IdG3sqgCEaRV3urPLA29IDudCufD89RYCohfBPNeWIM

This PR is copy of #67771 with resolved comments.

**Release note**:
```release-note
Replace scale down forbidden window with scale down stabilization window. Rather than waiting a fixed period of time between scale downs HPA now scales down to the highest recommendation it during the scale down stabilization window.
```

### CODE REVIEW

**Comment by idealhack** [comment]
@krzysztof-jastrzebski Hi, could you explain a bit about why there are adding instead of replacing in some places of this change?

**Comment by krzysztof-jastrzebski** [comment]
@idealhack I don't understand. Could you point in code what is added instead of being replaced?

**Comment by idealhack** [comment]
@krzysztof-jastrzebski It actually makes sense after I checked them again, `HorizontalPodAutoscalerDownscaleForbiddenWindow` is deprecated but not deleted for now.

**Comment by krzysztof-jastrzebski** [comment]
This PR is copy of https://github.com/kubernetes/kubernetes/pull/67771 with resolved comments.

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 1]
```diff
@@ -312,6 +312,8 @@ type HPAControllerConfiguration struct {
 	// horizontalPodAutoscalerUpscaleForbiddenWindow is a period after which next upscale allowed.
 	HorizontalPodAutoscalerUpscaleForbiddenWindow metav1.Duration
 	// horizontalPodAutoscalerDownscaleForbiddenWindow is a period after which next downscale allowed.
```
Comment (DirectXMan12) [inline pkg/apis/componentconfig/v1alpha1/types.go]: godoc comment is incorrect

**Review by DirectXMan12 — COMMENTED** [review]
minor godoc comment inline, otherwise looks good

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 1 above)
Comment (krzysztof-jastrzebski) [reply pkg/apis/componentconfig/v1alpha1/types.go]: Done.

**Comment by jbartosik** [comment]
Thanks for taking care of this :)

**Comment by jbartosik** [comment]
@thockin please approve API exceptions brought by adding new flags to HPA which need corresponding v1alpha1 componentconfig fields like the existing flags do (as I understand the exception is that the field names start with capital letters).

**Comment by thockin** [comment]
Is this aiming for 1.12 ?

**Comment by krzysztof-jastrzebski** [comment]
Yes, we would like to have it in 1.12.

**Comment by jbartosik** [comment]
@thockin yes, but please keep in mind that's an improvement basically from https://github.com/kubernetes/kubernetes/pull/67771 which @krzysztof-jastrzebski updated to address the issue @DirectXMan12 brought (while I don't have access to proper dev environment).

**Comment by jbartosik** [comment]
s/brought/addressed/

**Comment by DirectXMan12** [comment]
/hold
What prevents a bunch of updates to the object from making the window grow very large?  If I update my hpa once per second will we store 300 data points by accident?

**Comment by DirectXMan12** [comment]
More generally, the fact that I've just thought of this makes me nervous that there's something that I haven't thought of.  I'm hesitant about merging this this soon before code freeze -- ideally, I'd like a bit more soak time on this.  I'm not sure I want to unconditionally block this (especially since I may be incommunicado Monday morning, but it does make me nervous, and I'd prefer to leave this until next release when we have a bit more soak time.

**Comment by mwielgus** [comment]
@DirectXMan12 
We are not proposing increasing HPA frequency to 1 second. Actually, with the current design I would be against it. There is no way we can have, lets say 1000 HPAs, and still reconcile them all every 1 second, especially given the sing-threaded nature of the control loop and the need to call MetricsServer. 

Please let us know if it clarifies the issue.

**Comment by DirectXMan12** [comment]
@mweiglus we trigger reconciles every time the user updates the HPA due to how the informers work, IIRC.

**Comment by mwielgus** [comment]
@DirectXMan12 
It seems quite unlikely that the user will will update a significant number of HPAs every 1 second, and if he does, we will have a much bigger problem than 8kb consumed for additional samples. 

Moreover, currently the performance of HPA evaluation is relatively low (like 3-4 hpas/s) due  to:

* Single thread used to query Metrics Server
* Not using informer for listing pods (@krzysztof-jastrzebski is filing a bug as we speak).

So most of these large number of hpas will be sitting in a queue waiting waiting to be processed and additional updates to them will not change anything.

**Comment by DirectXMan12** [comment]
@krzysztof-jastrzebski ah, you're right, I'd forgotten that I fixed that (although you didn't link to the right line).

I'm going to unhold this since @mwielgus is on board, but I'd really still prefer not to merge things like this this late in code slush.

/unhold

**Pull request merged**